drivers/segger/serial_rtt.c

```diff
@@ -0,0 +1,380 @@
+
+static void serial_rtt_txint(FAR struct uart_dev_s *dev, bool enable)
+{
+}
+
+/****************************************************************************
+ * Name: serial_rtt_txready
+ ****************************************************************************/
+
+static bool serial_rtt_txready(FAR struct uart_dev_s *dev)
+{
+    FAR struct serial_rtt_dev_s *rtt =
+             (FAR struct serial_rtt_dev_s *)dev->priv;
+  return SEGGER_RTT_GetAvailWriteSpace(rtt->channel) != 0;
+}
+
+/****************************************************************************
+ * Name: serial_rtt_txempty
+ ****************************************************************************/
+
+static bool serial_rtt_txempty(FAR struct uart_dev_s *dev)
+{
+  FAR struct serial_rtt_dev_s *rtt =
+             (FAR struct serial_rtt_dev_s *)dev->priv;
+  return SEGGER_RTT_GetBytesInBuffer(rtt->channel) == 0;
+}
+
+/****************************************************************************
+ * Name: serial_rtt_loop
+ ****************************************************************************/
+
+static int serial_rtt_loop(int argc, FAR char *argv[])
+{
+  while (1)
+    {
+      FAR struct serial_rtt_dev_s *rtt;
+      irqstate_t flags;
+
+      nxsig_usleep(CONFIG_SERIAL_RTT_POLLING_INTERVAL);
+      flags = spin_lock_irqsave(NULL);
+      list_for_every_entry(&g_serial_rtt_list,
+                           rtt, struct serial_rtt_dev_s, node)
+        {
+          spin_unlock_irqrestore(NULL, flags);
+
+          /* Check if there is data update */
+
+          serial_rtt_dmarxfree(&rtt->uart);
+          flags = spin_lock_irqsave(NULL);
+        }
+
+      spin_unlock_irqrestore(NULL, flags);
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+void serial_rtt_register(FAR const char *name, int channel,
+                         bool isconsole, size_t txsize, size_t rxsize)
+{
+  irqstate_t flags;
+  FAR struct serial_rtt_dev_s *rtt =
+             kmm_zalloc(sizeof(struct serial_rtt_dev_s));
+  rtt->channel = channel;
+  rtt->uart.isconsole = isconsole;
+  rtt->uart.ops = &g_serial_rtt_ops;
+  rtt->uart.priv = rtt;
+  rtt->uart.xmit.size = txsize;
+  rtt->uart.xmit.buffer = kmm_zalloc(txsize);
+  DEBUGASSERT(rtt->uart.xmit.buffer);
+  rtt->uart.recv.size = rxsize;
+  rtt->uart.recv.buffer = kmm_zalloc(rxsize);
+  DEBUGASSERT(rtt->uart.recv.buffer);
+
+  if (channel != 0)
+    {
+      SEGGER_RTT_ConfigUpBuffer(channel, name, rtt->uart.xmit.buffer,
+                                txsize, SEGGER_RTT_MODE_DEFAULT);
+      SEGGER_RTT_ConfigDownBuffer(channel, name, rtt->uart.recv.buffer,
+                                  rxsize, SEGGER_RTT_MODE_DEFAULT);
+    }
+
+  if (list_is_empty(&g_serial_rtt_list))
+    {
+      kthread_create("serial_rtt_daemon", SCHED_PRIORITY_MAX,
```

Comment:
should we use high priority work quue to save a thread

Reply:
ping @Gary-Hobson